src/Authentication/Authentication/Common/GraphSettings.cs

```diff
@@ -0,0 +1,312 @@
+        public void Save(string path)
+        {
+            if (!string.IsNullOrEmpty(path))
+            {
+                using (var provider = ProtectedFileProvider.CreateFileProvider(path, FileProtection.ExclusiveWrite))
+                {
+                    Save(provider);
+                }
+            }
+        }
+
+        public void Save(IFileProvider provider)
+        {
+            // Remove built-in environments.
+            foreach (string env in GraphEnvironment.BuiltInEnvironments.Keys)
+            {
+                EnvironmentTable.Remove(env);
+            }
+
+            try
+            {
+                string contents = ToString();
+                string diskContents = string.Empty;
+                diskContents = provider.CreateReader().ReadToEnd();
+
+                if (diskContents != contents)
+                {
+                    var writer = provider.CreateWriter();
+                    writer.Write(contents);
+                    writer.Flush();
+
+                    // When writing to a stream, ensure that the file is truncated
+                    // so that previous data is overwritten
+                    provider.Stream.SetLength(provider.Stream.Position);
+                }
+            }
+            finally
+            {
+                // Add back built-in environments.
+                foreach (GraphEnvironment env in GraphEnvironment.BuiltInEnvironments.Values)
+                {
+                    EnvironmentTable[env.Name] = env;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Safely deserializes JSON string to an object.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="serialization">A string to deserialize.</param>
+        /// <param name="result">An object with deserialized content.</param>
+        /// <param name="converter">A <see cref="JsonConverter"/> to use to deserialize JSON content.</param>
+        /// <returns>True is successful, otherwise false.</returns>
+        internal bool TryDeserializeObject<T>(string serialization, out T result, JsonConverter converter = null)
+        {
+            result = default(T);
+            bool success = false;
+            try
+            {
+                result = converter == null ? JsonConvert.DeserializeObject<T>(serialization) : JsonConvert.DeserializeObject<T>(serialization, converter);
+                success = true;
+            }
+            catch (JsonException)
+            {
+            }
+
+            return success;
+        }
+
+        /// <summary>
+        /// Gets a list of <see cref="IGraphEnvironment"/>.
+        /// </summary>
+        /// <param name="name">The name of the environment to remove.</param>
+        /// <returns>A list of <see cref="IGraphEnvironment"/></returns>
+        public List<IGraphEnvironment> ListEnvironments(string name)
```

Comment:
should this be default so ListEnvironments just adds the range?

Reply:
This isn't clear. ListEnvironments serves 3 purposes:
- When name is not provided, it returns a sorted list of all registered environments.
- When an environment name is specified and it exists in the settings, that environment object will be returned.
- When an environment name is specified and it doesn't exists in the settings, an empty list returned.